### ISSUE TITLE
Error installingGitLab

### ISSUE THREAD

**Issue by Plumeros** [YunoHost-Apps/gitlab_ynh#229]
### Describe the bug

Error installing GitLab app in yunohost.

### Context

Issue found on Odroid C4 as well as on a amd64 virtual machine. Both with installed Debian Bullseye before yunohost installation and config.
The only manually app installed is Nextcloud.
Details:
- YunoHost version: latest, updated today 2023-09-06
- I have access to my server: direct access for the VM, ssh and webadmin for the Odroid C4
- Are you in a special context or did you perform some particular tweaking on your YunoHost instance?: no
- Using, or trying to install package version/branch: Gitlab, latest 2023-09-06

### Steps to reproduce

1.Install Debian Bullseye
2.Install yunohost
3. Install Nextcloud
4. Install Gitlab

### Expected behavior

Gitlab is installed

### Logs

https://paste.yunohost.org/raw/ebozisatag

**Comment by kay0u** [comment]
Hi,

At the end of logs there is:

```
wget --tries 3 --no-dns-cache --timeout 900 --no-verbose --output-document=/var/cache/yunohost/download/gitlab/x86-64 https://packages.gitlab.com/gitlab/gitlab-ce/packages/debian/bullseye/gitlab-ce_16.3.0-ce.0_amd64.deb/download.deb
2023-09-06 07:52:09,422: DEBUG - + out='fehlgeschlagen: Die Wartezeit für die Verbindung ist abgelaufen.
2023-09-06 07:52:09,422: DEBUG - fehlgeschlagen: Die Wartezeit für die Verbindung ist abgelaufen.
2023-09-06 07:52:09,423: DEBUG - fehlgeschlagen: Die Wartezeit für die Verbindung ist abgelaufen.
2023-09-06 07:52:09,423: DEBUG - fehlgeschlagen: Die Wartezeit für die Verbindung ist abgelaufen.
2023-09-06 07:52:09,423: DEBUG - fehlgeschlagen: Die Wartezeit für die Verbindung ist abgelaufen.
2023-09-06 07:52:09,424: DEBUG - fehlgeschlagen: Die Wartezeit für die Verbindung ist abgelaufen.
2023-09-06 07:52:09,424: DEBUG - fehlgeschlagen: Die Wartezeit für die Verbindung ist abgelaufen.
2023-09-06 07:52:09,424: DEBUG - fehlgeschlagen: Die Wartezeit für die Verbindung ist abgelaufen.
2023-09-06 07:52:09,425: DEBUG - fehlgeschlagen: Die Wartezeit für die Verbindung ist abgelaufen.
2023-09-06 07:52:09,425: DEBUG - fehlgeschlagen: Keine Route zum Zielrechner.
[...]
```

I don't speak German, but my translator does:

```
Failed: The waiting time for the connection has expired.
failed: No route to the destination computer.
```

I'm not sure why you have this error, maybe a DNS problem or an internet issue? You can either retry to installing it again or, try to pre-downloading [the file](https://packages.gitlab.com/gitlab/gitlab-ce/packages/debian/bullseye/gitlab-ce_16.3.0-ce.0_amd64.deb) and placing it in the folder (create it if necessary) `/opt/yunohost-apps-src/gitlab/x86-64` (with the name gitlab-ce_16.3.0-ce.0_amd64.deb.

After that you can restart the install and remove the file at the end of the installation.

**Comment by Plumeros** [comment]
Hi,
thanks for the answer. I just re-tried the installation which is easy on a virtual machine. The network issue persists. The strange thing is that generally the VM has internet access, but downloading the package seems to be an issue. I will try the trick with pre-loading the package.
Cheers

**Comment by Plumeros** [comment]
The installation proceeded further. It was able to copy the pre-loaded file. But then the next issue occurs.
See also
https://paste.yunohost.org/raw/olazewoles

2023-09-08 21:08:00,686: WARNING - cp: -r wurde nicht angegeben, daher wird das Verzeichnis '/opt/yunohost-apps-src/gitlab/x86-64' ausgelassen
2023-09-08 21:08:00,687: DEBUG - + ynh_exit_properly
2023-09-08 21:08:00,688: DEBUG - + [[ install =~ ^install$|^upgrade$|^restore$ ]]
2023-09-08 21:08:02,197: ERROR - Unable to install gitlab: An error occurred inside the app installation script

The german part means that cp: -r  was not set, therefore '/opt/yunohost-apps-src/gitlab/x86-64' was omitted.

I think there is a * missing. The cp command should be
cp /opt/yunohost-apps-src/gitlab/x86-64/* /var/cache/yunohost/download/gitlab/x86-64

But in which script?

Cheers

**Comment by kay0u** [comment]
`x86-64` should be ne pre-loaded file, not a folder where you put the file

**Comment by Plumeros** [comment]
Ah, with the downloaded file called x86-64 in the folder /opt/yunohost-apps-src/gitlab, GitLab could be installed in the virtual machine.

The strange thing is that calling 
wget https://packages.gitlab.com/gitlab/gitlab-ce/packages/debian/bullseye/gitlab-ce_16.3.0-ce.0_amd64.deb/download.deb
in  the yunohost server runs into a timeout. Getting a file from my own server works fine. Getting the deb package on my debian PC works fine too.

Now I would like to do the same trick on my Odroid C4 which runs on Armbian.

What should be the name of the file in /opt/yunohost-apps-src/gitlab GitLab in this case?

Thanks and kind regards

**Comment by kay0u** [comment]
I can't say what's the source of the problem :/ but the problem is that as soon as gitlab is updated, you'll have to do the same thing.

wget in Yunohost server has a 900s timeout, do you have a slow network?

 > Now I would like to do the same trick on my Odroid C4 which runs on Armbian.
 
 Tell me if I'm wrong, but this is an ARM 64 bits processor? I'd say `arm64`, and download the file
https://packages.gitlab.com/gitlab/gitlab-ce/packages/debian/bullseye/gitlab-ce_16.3.0-ce.0_arm64.deb/download.deb

**Comment by Plumeros** [comment]
Yes, it's an ARM64 processor, thanks for the hint. But for having a sustainable solution I need to find the root cause of the timeout issue. The network is generally fast, but maybe it's an issue with my home network config.

Meanwhile I tried to install git on the ARM64 server without the pre-load trick. There was a warning about the missing swap file, which may cause issues with gitlab. The installation prodeeded, but was then stuck a few steps later (I don't know exactly where) and the web UI was not responsive anymore. I had to connect via ssh for shutting down the server.
For luck I created a 1:1 copy of the SD card before the git installation, so I exchanged the card and restarted the server with the healthy card.

Now I have the SD card with the failed gitlab installation on it. Maybe I can find some traces in the file system. Can you tell me where to find the log files? Or do you have a link to some documentation describing this?

Thanks and kind regards

**Comment by kay0u** [comment]
Gitlab can takes time to install. You have to be patient.

You can find logs in `/var/logs/yunohost/yunohost-api` (if you install it from webui) or `cli` if you install from cmd

If your server is stuck during the install you can `tail -f` these files to ensure that the server is still running correctly

**Comment by Plumeros** [comment]
I was waiting for about 4h, I think the installation should not take that long. Thanks for the hints, I will follow them as soon as time permits.
Additionally I try to find the root cause of these network issues, as the rest of the machines generally do not have connectivity issues.
Thanks for helping anyway :)

**Comment by Plumeros** [comment]
Now I'm a few steps further.

The network issue was related to a virtualbox bridging issue with the wifi adapter. Once I switched to the lan adapter, the installation worked like a charm.

But when I try to install Gitlab on the Odroid C4, I run into the same issue as before, but with some more details:

The ssh terminal is still responsive, but the web ui is frozen now, showing:
Storing installation settings...
Adding 156 Mo to swap...
The main mountpoint of your system '/' is on an SD card, swap will not be added to prevent some damage of this one, but that can cause troubles for the app gitlab. If you still want activate the swap, you can relaunch the command preceded by 'SD_CARD_CAN_SWAP=1'
Preconfigure GitLab...
Setting up source files...

tail -f /var/log/yunohost/yunohost-api.log looks good until:
2023-09-15 18:18:01,056 DEBUG    yunohost.hook (unknown function) - [1311.10]       * template[/opt/gitlab/sv/logrotate/finish] action create (skipped due to only_if)
2023-09-15 18:18:01,068 DEBUG    yunohost.hook (unknown function) - [1311.10]       * directory[/opt/gitlab/sv/logrotate/control] action create (up to date)
2023-09-15 18:18:01,152 DEBUG    yunohost.hook (unknown function) - [1311.10]       * template[/opt/gitlab/sv/logrotate/control/t] action create (up to date)
2023-09-15 18:18:01,157 DEBUG    yunohost.hook (unknown function) - [1311.10]       * link[/opt/gitlab/init/logrotate] action create (up to date)
2023-09-15 18:18:01,159 DEBUG    yunohost.hook (unknown function) - [1311.10]       * file[/opt/gitlab/sv/logrotate/down] action nothing (skipped due to action :nothing)
2023-09-15 18:18:01,173 DEBUG    yunohost.hook (unknown function) - [1311.10]       * directory[/opt/gitlab/service] action create (up to date)
2023-09-15 18:18:01,178 DEBUG    yunohost.hook (unknown function) - [1311.10]       * link[/opt/gitlab/service/logrotate] action create[2023-09-15T18:18:01+02:00] INFO: link[/opt/gitlab/service/logrotate] created
2023-09-15 18:18:01,180 DEBUG    yunohost.hook (unknown function) - [1311.10] 
2023-09-15 18:18:01,181 DEBUG    yunohost.hook (unknown function) - [1311.10]         - create symlink at /opt/gitlab/service/logrotate to /opt/gitlab/sv/logrotate

This is the last entry, no further entries are added since 1/2h now.

As there was a warning regarding the swap, below you can find some additional information:

free -h:
               total        used        free      shared  buff/cache   available
Mem:           3.7Gi       782Mi        56Mi       1.2Gi       2.9Gi       1.6Gi
Swap:          1.8Gi        27Mi       1.8Gi

swapon -s:
Filename				Type		Size	Used	Priority
/dev/zram0                             	partition	1938144	28160	5

I think that armbian is using zram as swap partition. I'm not sure if the installation can handle that. In the yunohost-api.log there are a couple of "swap" entries:
[gitlab_tail.log](https://github.com/YunoHost-Apps/gitlab_ynh/files/12621994/gitlab_tail.log)

I hope this info is useful and that it's not too much burden for you to have a look into.
Thanks and kind regards

**Comment by Plumeros** [comment]
Btw.: the memory onthe SD card should not be an issue:

df -h:
Filesystem      Size  Used Avail Use% Mounted on
udev            1.8G     0  1.8G   0% /dev
tmpfs           379M  5.6M  374M   2% /run
/dev/mmcblk0p1   29G   18G   11G  62% /
tmpfs           1.9G   28K  1.9G   1% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           1.9G  1.2G  691M  64% /tmp
/dev/zram1       49M   31M   15M  69% /var/log
tmpfs           379M     0  379M   0% /run/user/15526

**Comment by Plumeros** [comment]
And finally I found the code format. Voilà:

```
free -h
               total        used        free      shared  buff/cache   available
Mem:           3.7Gi       782Mi        56Mi       1.2Gi       2.9Gi       1.6Gi
Swap:          1.8Gi        27Mi       1.8Gi

swapon -s
Filename				Type		Size	Used	Priority
/dev/zram0                             	partition	1938144	28160	5

df -h
Filesystem      Size  Used Avail Use% Mounted on
udev            1.8G     0  1.8G   0% /dev
tmpfs           379M  5.6M  374M   2% /run
/dev/mmcblk0p1   29G   18G   11G  62% /
tmpfs           1.9G   28K  1.9G   1% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           1.9G  1.2G  691M  64% /tmp
/dev/zram1       49M   31M   15M  69% /var/log
tmpfs           379M     0  379M   0% /run/user/15526
```

**Comment by Plumeros** [comment]
Some news on this issue :)

I have set up an additional Odroid C4 as an internal test yunohost and tried to install gitlab as the only app (no other apps installed) and I didn't change any configuration, e.g. zram.
The log entries look different from the trial on my productive Odroid C4, but unfortunately the installation still failed.

The error seems to happen after:
2023-10-21 11:52:37,176
see yunopaste:
https://paste.yunohost.org/raw/kawiciweva

Does gitlab work on any other SBC, e.g. the Raspi 4?

**Comment by Plumeros** [comment]
OK, I'm able to answer the last question by myself now. After switching my SBC from Odroid C4 4GB to a Raspi 4B 8GB, the installation and configuration of Gitlab worked like a charm. I have already ported my first projects, all good.
Hence if one takes a Raspi 4B with 8GB RAM and the yunohost image, Gitlab should work.
The question why it didn't work on the initial setup with an Odroid C4 and Armbian underlaying yunohost remains open.
One difference I found is: Armbian has zram installed, where the Raspi has just a 100MB swap file.
If I have the time, I may investigate on that.
